### PR TITLE
support @Decorated bean injection in decorator

### LINKED ISSUE #43863: Quarkus does not support injecting @Decorated in decorator
### Describe the bug

per CDI spec:

> 
> the container must provide beans allowing decorators to obtain information about the beans they intercept and decorate:
> 
>     a bean with scope @Dependent, qualifier @Decorated and type Bean which can be injected into any decorator instance.

This is not supported by quarkus yet.

### PR DESCRIPTION
**Pull request by rmanibus**
- Fixes: #43863

### CODE REVIEW

**Comment by rmanibus** [comment]
cc @mkouba , @Ladicek

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BuiltinBean.java [hunk 1]
```diff
@@ -515,6 +534,13 @@ private static void validateInterceptedBean(ValidatorContext ctx) {
         }
     }
 
+    private static void validateDecoratedBean(ValidatorContext ctx) {
+        if (ctx.injectionTarget.kind() != InjectionTargetInfo.TargetKind.BEAN
+                || !ctx.injectionTarget.asBean().isDecorator()) {
+            ctx.errors.accept(new DefinitionException("Only interceptors can access intercepted bean metadata"));
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BuiltinBean.java:540]:
```suggestion
            ctx.errors.accept(new DefinitionException("Only decorators can access decorated bean metadata"));
```

File: independent-projects/vertx-utils/src/main/java/io/quarkus/vertx/utils/VertxOutputStream.java [hunk 2]
```diff
@@ -73,7 +73,6 @@ public void handle(AsyncResult<Void> event) {
         });
     }
 
-
```
Comment (mkouba) [inline independent-projects/vertx-utils/src/main/java/io/quarkus/vertx/utils/VertxOutputStream.java:76]: ??

**Comment by Ladicek** [comment]
Some validations are missing, but otherwise LGTM on the first sight.

File: independent-projects/vertx-utils/src/main/java/io/quarkus/vertx/utils/VertxOutputStream.java (hunk 2 above)
Comment (rmanibus) [reply independent-projects/vertx-utils/src/main/java/io/quarkus/vertx/utils/VertxOutputStream.java]: oups, rolled back !

**Comment by rmanibus** [comment]
I think I fixed everything. The build should pass now

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BuiltinBean.java [hunk 3]
```diff
@@ -321,6 +327,19 @@ private static void generateInterceptedBeanBytecode(GeneratorContext ctx) {
...
+
+        ResultHandle decoratedBeanMetadataProviderSupplier = ctx.constructor.newInstance(
+                MethodDescriptors.FIXED_VALUE_SUPPLIER_CONSTRUCTOR, decoratedBeanMetadataProvider);
+        ctx.constructor.writeInstanceField(
+                FieldDescriptor.of(ctx.clazzCreator.getClassName(), ctx.providerName,
+                        Supplier.class.getName()),
+                ctx.constructor.getThis(),
+                decoratedBeanMetadataProviderSupplier);
+    }
+
```
Comment (Ladicek) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BuiltinBean.java:342]: This method is basically a copy of `generateInterceptedBeanBytecode`, and the class `DecoratedBeanMetadataProvider` is basically a copy of `InterceptedBeanMetadataProvider`. I was wondering if we could merge them, but it's probably not a big deal if they stay as they are.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BuiltinBean.java [hunk 4]
```diff
@@ -515,6 +534,13 @@ private static void validateInterceptedBean(ValidatorContext ctx) {
         }
     }
 
+    private static void validateDecoratedBean(ValidatorContext ctx) {
+        if (ctx.injectionTarget.kind() != InjectionTargetInfo.TargetKind.BEAN
+                || !ctx.injectionTarget.asBean().isDecorator()) {
+            ctx.errors.accept(new DefinitionException("Only decorators can access intercepted bean metadata"));
```
Comment (Ladicek) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BuiltinBean.java:540]:
```suggestion
            ctx.errors.accept(new DefinitionException("Only decorators can access decorated bean metadata"));
```

**Comment by Ladicek** [comment]
As mentioned above, some validations are missing. I implemented them in this branch: https://github.com/Ladicek/quarkus/commits/arc-decorated-validations/ Your commit from this PR is unchanged, I added a second commit on top of that.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/DecoratedBeanMetadataProvider.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import jakarta.enterprise.context.spi.CreationalContext;
+import jakarta.enterprise.inject.Decorated;
+import jakarta.enterprise.inject.spi.Bean;
+
+import io.quarkus.arc.InjectableReferenceProvider;
+
+/**
+ * {@link Decorated} {@link Bean} metadata provider.
+ */
+public class DecoratedBeanMetadataProvider implements InjectableReferenceProvider<Contextual<?>> {
```
Comment (manovotn) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/DecoratedBeanMetadataProvider.java:15]: As this is a copy of `InterceptedBeanMetadataProvider`, I'd suggest renaming the original class to something like `InterceptedDecoratedBeanMetadataProvider` and adjusting comments in it to explain it is used for both.
With that, you can also get rid of the extra method `BuiltinBean#generateDecoratedBeanBytecode` and perform similar change for the original method.

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/decorated/InvalidDecoratedTest.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+                "Invalid injection of @Decorated Bean<T>, can only be injected into decorators but was detected in: "
+                        + InvalidBean.class.getName() + "#decorated"));
+    }
+
+    @ApplicationScoped
+    static class InvalidBean {
+
+        @Inject
+        @Decorated
+        Bean<?> decorated;
```
Comment (manovotn) [inline independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/decorated/InvalidDecoratedTest.java:37]: This is actually not a valid injection point - there has to be a type corresponding to the delegate type, [see this section of CDI spec](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0#bean_metadata_full).
This means some form of validation is missing.

Comment (manovotn) [reply independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/decorated/InvalidDecoratedTest.java:37]: Ah, I didn't notice Ladislav beat me to it with his comment - https://github.com/quarkusio/quarkus/pull/43864#issuecomment-2413148607
Just apply his changes to add the validation :)

Comment (rmanibus) [reply independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/decorated/InvalidDecoratedTest.java]: just merged it ! thanks a lot !

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/DecoratedBeanMetadataProvider.java (hunk 5 above)
Comment (rmanibus) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/DecoratedBeanMetadataProvider.java]: @manovotn it's done, should I squash everything ?

**Review by Ladicek — APPROVED** [review]

**Review by manovotn — APPROVED** [review]

File: docs/src/main/asciidoc/cdi.adoc [hunk 7]
```diff
@@ -470,8 +475,9 @@ public class LargeTxAccount implements Account { <3>
 <2> `@Decorator` marks a decorator component.
 <3> The set of decorated types includes all bean types which are Java interfaces, except for `java.io.Serializable`.
 <4> Each decorator must declare exactly one _delegate injection point_. The decorator applies to beans that are assignable to this delegate injection point.
-<5> Decorators can inject other beans.
-<6> The decorator may invoke any method of the delegate object. And the container invokes either the next decorator in the chain or the business method of the intercepted instance.
+<5> It is possible to get contextual info about the decorated bean by using the @Decorated qualifier.
```
Comment (mkouba) [inline docs/src/main/asciidoc/cdi.adoc:478]:
```suggestion
<5> It is possible to obtain information about the decorated bean by using the `@Decorated` qualifier.
```

**Review by mkouba — COMMENTED** [review]
Could you pls squash the commits?

**Comment by rmanibus** [comment]
@mkouba sure thing, it's done !

**Review by mkouba — APPROVED** [review]

**Pull request merged**